src/custom/utils/operator/errors/QuoteError.ts

```diff
@@ -0,0 +1,78 @@
+import { ApiErrorCodes } from './OperatorError'
+
+export interface QuoteErrorObject {
+  errorType: QuoteErrorCodes
+  description: string
+}
+
+// Conforms to backend API
+// https://github.com/gnosis/gp-v2-services/blob/0bd5f7743bebaa5acd3be13e35ede2326a096f14/orderbook/openapi.yml#L562
+export enum QuoteErrorCodes {
+  InsufficientLiquidity = 'InsufficientLiquidity',
+  FeeExceedsFrom = 'FeeExceedsFrom',
+  UNHANDLED_ERROR = 'UNHANDLED_ERROR'
+}
+
+export enum QuoteErrorDetails {
+  InsufficientLiquidity = 'Token pair selected has insufficient liquidity',
+  FeeExceedsFrom = 'Current fee exceeds input "from" amount',
+  UNHANDLED_ERROR = 'An unknown error occurred while fetching the quote'
+}
+
+export function mapOperatorErrorToQuoteError(errorType: ApiErrorCodes): QuoteErrorObject {
+  switch (errorType) {
+    case ApiErrorCodes.NotFound:
+      return { errorType: QuoteErrorCodes.InsufficientLiquidity, description: QuoteErrorDetails.InsufficientLiquidity }
+    default:
+      return { errorType: QuoteErrorCodes.UNHANDLED_ERROR, description: QuoteErrorDetails.UNHANDLED_ERROR }
+  }
+}
+
+export default class QuoteError extends Error {
+  name = 'QuoteErrorObject'
+  type: QuoteErrorCodes
+  description: QuoteErrorObject['description']
+
+  // Status 400 errors
+  // https://github.com/gnosis/gp-v2-services/blob/9014ae55412a356e46343e051aefeb683cc69c41/orderbook/openapi.yml#L563
+  static quoteErrorDetails = QuoteErrorDetails
+
+  static async getErrorMessage(response: Response) {
+    try {
+      const orderPostError: QuoteErrorObject = await response.json()
+
+      if (orderPostError.errorType) {
+        return QuoteError.quoteErrorDetails[orderPostError.errorType]
+      } else {
+        console.error('Unknown reason for bad quote fetch', orderPostError)
+        return orderPostError.description
+      }
+    } catch (error) {
+      console.error('Error handling 400/404 error. Likely a problem deserialising the JSON response')
+      return QuoteError.quoteErrorDetails.UNHANDLED_ERROR
+    }
+  }
+
+  static async getErrorFromStatusCode(response: Response) {
+    switch (response.status) {
+      case 400:
+      case 404:
+        return this.getErrorMessage(response)
+
+      case 500:
+      default:
+        return 'Error fetching quote'
```

Comment:
I think in this unexpected case, we should at least `console.error` something. It would mean that the backend returns a error code we don't know about

Reply:
👍

Reply:
handled here: 98a200a